File: python/apps/connect4.py
from machine import Pin
from neopixel import NeoPixel
from time import sleep_ms, ticks_ms, ticks_diff
import _screens as screens

NAME = "Connect4"
NUM_LEDS = 64

COLS = 8
ROWS = 7  # LED rows 0..6 hold pieces; LED row 7 is the cursor lane.

RED = 1
YELLOW = 2

# Design hexes scaled to the project's 25% brightness convention:
# placed pieces use #ff0000 / #ffff00; the dim cursor mirrors the design's #800000.
PLACED = {RED: (64, 0, 0),  YELLOW: (64, 64, 0)}
CURSOR = {RED: (32, 0, 0),  YELLOW: (32, 32, 0)}

FRAME_MS = 50
FALL_STEP_MS = 70
PULSE_CYCLES = 3
PULSE_HALF_MS = 220
DRAW_HOLD_MS = 1200

np = None
JOY_LEFT = None
JOY_RIGHT = None
JOY_CENTER = None


def clear():
    for i in range(NUM_LEDS):
        np[i] = (0, 0, 0)


def px(col, row, color):
    if 0 <= col <= 7 and 0 <= row <= 7:
        np[row * 8 + col] = color


def draw_pieces(board):
    for c in range(COLS):
        for r in range(ROWS):
            v = board[c][r]
            if v:
                px(c, r, PLACED[v])


def render_board(board, cursor_col, cursor_player):
    clear()
    draw_pieces(board)
    if cursor_player is not None:
        px(cursor_col, 7, CURSOR[cursor_player])
    np.write()


def column_top(board, col):
    """Lowest empty row in the column, or None if full."""
    for r in range(ROWS):
        if board[col][r] == 0:
            return r
    return None


def animate_drop(board, col, target_row, player):
    color = PLACED[player]
    for r in range(6, target_row - 1, -1):
        if screens.check_exit():
            return False
        clear()
        draw_pieces(board)
        px(col, r, color)
        np.write()
        sleep_ms(FALL_STEP_MS)
    return True


def winning_line(board, col, row, player):
    for dc, dr in ((1, 0), (0, 1), (1, 1), (1, -1)):
        line = [(col, row)]
        c, r = col + dc, row + dr
        while 0 <= c < COLS and 0 <= r < ROWS and board[c][r] == player:
            line.append((c, r))
            c += dc
            r += dr
        c, r = col - dc, row - dr
        while 0 <= c < COLS and 0 <= r < ROWS and board[c][r] == player:
            line.append((c, r))
            c -= dc
            r -= dr
        if len(line) >= 4:
            return line
    return None


def pulse_line(board, line, player):
    bright = PLACED[player]
    dim = (bright[0] // 4, bright[1] // 4, bright[2] // 4)
    cells = set(line)
    for _ in range(PULSE_CYCLES):
        for color in (dim, bright):
            if screens.check_exit():
                return False
            clear()
            for c in range(COLS):
                for r in range(ROWS):
                    v = board[c][r]
                    if not v:
                        continue
                    px(c, r, color if (c, r) in cells else PLACED[v])
            np.write()
            sleep_ms(PULSE_HALF_MS)
    return True


def hold_with_exit(ms):
    """Sleep ms while polling check_exit. Returns True if exit was triggered."""
    t0 = ticks_ms()
    while ticks_diff(ticks_ms(), t0) < ms:
        if screens.check_exit():
            return True
        sleep_ms(15)
    return False


def play_one_round():
    board = [[0] * ROWS for _ in range(COLS)]
    cursor_col = 4
    player = RED
    moves = 0

    prev_left = False
    prev_right = False
    center_press_start = None

    while True:
        if screens.check_exit():
            return None

        cur_left = JOY_LEFT.value() == 0
        cur_right = JOY_RIGHT.value() == 0
        cur_center = JOY_CENTER.value() == 0

        if cur_left and not prev_left and cursor_col > 0:
            cursor_col -= 1
        if cur_right and not prev_right and cursor_col < COLS - 1:
            cursor_col += 1

        # Drop on release of a short center tap. Long holds belong to check_exit.
        if cur_center and center_press_start is None:
            center_press_start = ticks_ms()
        elif not cur_center and center_press_start is not None:
            held = ticks_diff(ticks_ms(), center_press_start)
            center_press_start = None
            if held < screens.EXIT_HOLD_MS:
                target = column_top(board, cursor_col)
                if target is not None:
                    if not animate_drop(board, cursor_col, target, player):
                        return None
                    board[cursor_col][target] = player
                    moves += 1

                    line = winning_line(board, cursor_col, target, player)
                    if line:
                        if not pulse_line(board, line, player):
                            return None
                        return player

                    if moves >= COLS * ROWS:
                        render_board(board, cursor_col, None)
                        if hold_with_exit(DRAW_HOLD_MS):
                            return None
                        return 0

                    player = YELLOW if player == RED else RED

        prev_left = cur_left
        prev_right = cur_right

        render_board(board, cursor_col, player)
        sleep_ms(FRAME_MS)


def run(neopixel, joystick):
    global np, JOY_LEFT, JOY_RIGHT, JOY_CENTER
    np = neopixel
    JOY_LEFT = joystick["left"]
    JOY_RIGHT = joystick["right"]
    JOY_CENTER = joystick["center"]
    screens.init(neopixel, joystick)
    while True:
        if screens.loading_screen() == "exit":
            return
        outcome = play_one_round()
        if outcome is None:
            return
        if screens.end_screen() == "exit":
            return


if __name__ == "__main__":
    _np = NeoPixel(Pin(19, Pin.OUT), NUM_LEDS)
    _joy = {
        "up":     Pin(3, Pin.IN),
        "down":   Pin(6, Pin.IN),
        "left":   Pin(7, Pin.IN),
        "right":  Pin(2, Pin.IN),
        "center": Pin(8, Pin.IN),
        "slide":  Pin(9, Pin.IN),
    }
    run(_np, _joy)

File: web/src/apps/connect4.ts
import type { NeoPixel, RGB } from "../hardware/neopixel";
import type { Joystick, Pin } from "../hardware/joystick";
import * as screens from "../screens";
import { sleep_ms, ticks_diff, ticks_ms } from "../runtime/time";

export const NAME = "Connect4";

const NUM_LEDS = 64;
const COLS = 8;
const ROWS = 7; // LED rows 0..6 hold pieces; LED row 7 is the cursor lane.

const RED = 1;
const YELLOW = 2;
type Player = typeof RED | typeof YELLOW;

const PLACED: Record<Player, RGB> = {
  [RED]: [64, 0, 0],
  [YELLOW]: [64, 64, 0],
};
const CURSOR: Record<Player, RGB> = {
  [RED]: [32, 0, 0],
  [YELLOW]: [32, 32, 0],
};

const FRAME_MS = 50;
const FALL_STEP_MS = 70;
const PULSE_CYCLES = 3;
const PULSE_HALF_MS = 220;
const DRAW_HOLD_MS = 1200;

let np: NeoPixel;
let JOY_LEFT: Pin;
let JOY_RIGHT: Pin;
let JOY_CENTER: Pin;

function clear(): void {
  for (let i = 0; i < NUM_LEDS; i++) np[i] = [0, 0, 0];
}

function px(col: number, row: number, color: RGB): void {
  if (col >= 0 && col <= 7 && row >= 0 && row <= 7) {
    np[row * 8 + col] = color;
  }
}

type Board = number[][]; // board[col][row]; 0 = empty.

function makeBoard(): Board {
  const b: Board = [];
  for (let c = 0; c < COLS; c++) {
    const col: number[] = [];
    for (let r = 0; r < ROWS; r++) col.push(0);
    b.push(col);
  }
  return b;
}

function drawPieces(board: Board): void {
  for (let c = 0; c < COLS; c++) {
    for (let r = 0; r < ROWS; r++) {
      const v = board[c][r];
      if (v) px(c, r, PLACED[v as Player]);
    }
  }
}

function renderBoard(board: Board, cursorCol: number, cursorPlayer: Player | null): void {
  clear();
  drawPieces(board);
  if (cursorPlayer !== null) {
    px(cursorCol, 7, CURSOR[cursorPlayer]);
  }
  np.write();
}

function columnTop(board: Board, col: number): number | null {
  for (let r = 0; r < ROWS; r++) {
    if (board[col][r] === 0) return r;
  }
  return null;
}

async function animateDrop(board: Board, col: number, targetRow: number, player: Player): Promise<boolean> {
  const color = PLACED[player];
  for (let r = 6; r >= targetRow; r--) {
    if (screens.check_exit()) return false;
    clear();
    drawPieces(board);
    px(col, r, color);
    np.write();
    await sleep_ms(FALL_STEP_MS);
  }
  return true;
}

function winningLine(
  board: Board,
  col: number,
  row: number,
  player: Player,
): Array<[number, number]> | null {
  const dirs: ReadonlyArray<readonly [number, number]> = [
    [1, 0],
    [0, 1],
    [1, 1],
    [1, -1],
  ];
  for (const [dc, dr] of dirs) {
    const line: Array<[number, number]> = [[col, row]];
    let c = col + dc;
    let r = row + dr;
    while (c >= 0 && c < COLS && r >= 0 && r < ROWS && board[c][r] === player) {
      line.push([c, r]);
      c += dc;
      r += dr;
    }
    c = col - dc;
    r = row - dr;
    while (c >= 0 && c < COLS && r >= 0 && r < ROWS && board[c][r] === player) {
      line.push([c, r]);
      c -= dc;
      r -= dr;
    }
    if (line.length >= 4) return line;
  }
  return null;
}

async function pulseLine(board: Board, line: Array<[number, number]>, player: Player): Promise<boolean> {
  const bright = PLACED[player];
  const dim: RGB = [Math.floor(bright[0] / 4), Math.floor(bright[1] / 4), Math.floor(bright[2] / 4)];
  const cells = new Set(line.map(([c, r]) => `${c},${r}`));
  for (let i = 0; i < PULSE_CYCLES; i++) {
    for (const color of [dim, bright]) {
      if (screens.check_exit()) return false;
      clear();
      for (let c = 0; c < COLS; c++) {
        for (let r = 0; r < ROWS; r++) {
          const v = board[c][r];
          if (!v) continue;
          const onLine = cells.has(`${c},${r}`);
          px(c, r, onLine ? color : PLACED[v as Player]);
        }
      }
      np.write();
      await sleep_ms(PULSE_HALF_MS);
    }
  }
  return true;
}

async function holdWithExit(ms: number): Promise<boolean> {
  const t0 = ticks_ms();
  while (ticks_diff(ticks_ms(), t0) < ms) {
    if (screens.check_exit()) return true;
    await sleep_ms(15);
  }
  return false;
}

async function playOneRound(): Promise<Player | 0 | null> {
  const board = makeBoard();
  let cursorCol = 4;
  let player: Player = RED;
  let moves = 0;

  let prevLeft = false;
  let prevRight = false;
  let centerPressStart: number | null = null;

  while (true) {
    if (screens.check_exit()) return null;

    const curLeft = JOY_LEFT.value() === 0;
    const curRight = JOY_RIGHT.value() === 0;
    const curCenter = JOY_CENTER.value() === 0;

    if (curLeft && !prevLeft && cursorCol > 0) cursorCol -= 1;
    if (curRight && !prevRight && cursorCol < COLS - 1) cursorCol += 1;

    if (curCenter && centerPressStart === null) {
      centerPressStart = ticks_ms();
    } else if (!curCenter && centerPressStart !== null) {
      const held = ticks_diff(ticks_ms(), centerPressStart);
      centerPressStart = null;
      if (held < screens.EXIT_HOLD_MS) {
        const target = columnTop(board, cursorCol);
        if (target !== null) {
          if (!(await animateDrop(board, cursorCol, target, player))) return null;
          board[cursorCol][target] = player;
          moves += 1;

          const line = winningLine(board, cursorCol, target, player);
          if (line) {
            if (!(await pulseLine(board, line, player))) return null;
            return player;
          }

          if (moves >= COLS * ROWS) {
            renderBoard(board, cursorCol, null);
            if (await holdWithExit(DRAW_HOLD_MS)) return null;
            return 0;
          }

          player = player === RED ? YELLOW : RED;
        }
      }
    }

    prevLeft = curLeft;
    prevRight = curRight;

    renderBoard(board, cursorCol, player);
    await sleep_ms(FRAME_MS);
  }
}

export async function run(neopixel: NeoPixel, joystick: Joystick): Promise<void> {
  np = neopixel;
  JOY_LEFT = joystick.left;
  JOY_RIGHT = joystick.right;
  JOY_CENTER = joystick.center;
  screens.init(neopixel, joystick);
  while (true) {
    if ((await screens.loading_screen()) === "exit") return;
    const outcome = await playOneRound();
    if (outcome === null) return;
    if ((await screens.end_screen()) === "exit") return;
  }
}
